### ISSUE TITLE
Xiaomi Magic Cube - XMMF01JQD

### ISSUE THREAD

**Issue by andrewjswan** [custom-components/ble_monitor#271]
I will make a separate thread, add information, if you can add good, no so can someone add and then there will be more information.

In https://github.com/AlexxIT/XiaomiGateway3 Readme - Xiaomi Magic Cube (XMMF01JQD) - don't sends edge info, only direction!
In source code, description:  https://github.com/AlexxIT/XiaomiGateway3/blob/829f748fc838da2d5f9a506371cfb446c89598ab/custom_components/xiaomi_gateway3/core/bluetooth.py#L14 and action: https://github.com/AlexxIT/XiaomiGateway3/blob/829f748fc838da2d5f9a506371cfb446c89598ab/custom_components/xiaomi_gateway3/core/bluetooth.py#L144

Additional info: https://medium.com/@juananclaramunt/xiaomi-mi-smart-rubik-cube-ff5a22549f90

Some logs in - https://github.com/custom-components/ble_monitor/issues/268#issuecomment-776918304

**Comment by Ernst79** [comment]
I looked at the two messages in the log in #268, but they do not contain data. If you can create a new log with a connected cube, perhaps I can find more info. Based on your links, it should have a similar message as the switch.

**Comment by andrewjswan** [comment]
> Didn't realize it was a magic rubic cube.

Very funny thing, I really do not understand how he is supposed to determine whether the cube is now collected, or it is controlled by a smart home.

Can you add some switch to the GUI, that you can enable it, specify the MAC address (or leave blank) and integration would output all the data you need for debugging in the log file, say the level of Error. Regardless of whether it is a known device or not.

That would be easier, but in this case you have to restart Home Assistant, and it's not always convenient. 

Fluid sensor SJWS01LM has a button, when you click on it it sends a message to the gateway, it would be convenient to have the fact of pressing in Home Assistant.

**Comment by andrewjswan** [comment]
Tried again, twisting the facet to the right, to the left from time to time, then twisting the cube and twisting the other facets to the right and to the left. But most of all the white facet.

I also tried pushing on the leak sensor SJWS01LM - 54:EF:44:E0:45:51
[home-assistant.log.zip](https://github.com/custom-components/ble_monitor/files/5981684/home-assistant.log.zip)
.

**Comment by Ernst79** [comment]
The message by SJWS01LM is not logged, as it is not an unknown device anymore. You have to use hcidump for this device. 

`sudo hcidump --raw hci > dump.txt`

**Comment by andrewjswan** [comment]
> The message by SJWS01LM is not logged, as it is not an unknown device anymore. You have to use hcidump for this device.

I dont have console for my HASS.IO ...

**Comment by andrewjswan** [comment]
> The message by SJWS01LM is not logged, as it is not an unknown device anymore.

It would be nice to have some kind of switch to keep track of events from known and unknown devices, without restarting Home Assistant. :)

**Comment by Ernst79** [comment]
I know.. 

There isn't any data of your cube in the logs either. I don't see any `1695fe` advertisements (from xiaomi), only `16cdfd` from qingping. 

You can always add a logger line in the code, at line 648. 

    def parse_raw_message(self, data):
        _LOGGER.error("Data: %s", data.hex())
        """Parse the raw data."""

**Comment by andrewjswan** [comment]
> There isn't any data of your cube in the logs either. I don't see any `1695fe` advertisements (from xiaomi), only `16cdfd` from qingping.
> 
> You can always add a logger line in the code, at line 648.

I'll try again in the next few days.

**Comment by Ernst79** [comment]
@andrewjswan Any update on this? Did you had a change to collect some data? In the latest version, you have to add the LOGGER line in `__init__.py` in the ble_parser subfolder.

**Comment by andrewjswan** [comment]
The child has the cube, I try it from time to time, but there is no meaningful information in the log yet. I do not know why. But if it works, I'll let you know more.

**Comment by Ernst79** [comment]
It took a while, but I think the Magic Cube is added to BLE monitor. It only shows the direction (left/right), not the state of the edges. The edge info is only available when setting up a connection, which is out of the scope of BLE monitor. 

Please give it a try in 6.1.0-beta.

**Comment by andrewjswan** [comment]
> Please give it a try in 6.1.0-beta.

I install 6.1.0-Beta, add:
```
  - mac: 'E2:53:30:XX:XX:XX'
    name: 'Magic Cube'
```
But i dont see any devices or sensors for Magic Cube...

**Comment by Ernst79** [comment]
Hmm, I tried to add it based on the info from XiaomiGateway3. If it isn't working, I do need some advertisements to check. Can you try to collect some advertisements with `report_unknown: Xiaomi`? I think you will need to install the old version first (6.0.1)

**Comment by andrewjswan** [comment]
> report_unknown: Xiaomi

What level of logging should be?

**Comment by Ernst79** [comment]
info level

**Comment by andrewjswan** [comment]
But I'm not sure about the correctness of Mac's address.
[home-assistant.log](https://github.com/custom-components/ble_monitor/files/7595730/home-assistant.log)

**Comment by Ernst79** [comment]
I found some messages, but they all do not containing data. 

Example 
`0f1695fe3030e1046c54d3e63053e20d`

`3030` means that the payload does not contain data, only a message to say "I'm here"
`e104` is the indication for the cube. 

Did you turn/rotate the cube while the collection of the data?

**Comment by andrewjswan** [comment]
> Did you turn/rotate the cube while the collection of the data?

Yes, but i try again ...

**Comment by andrewjswan** [comment]
```
2021-11-24 16:15:12 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -84, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1048e54d3e63053e20d
2021-11-24 16:15:14 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -83, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1048f54d3e63053e20d
2021-11-24 16:15:20 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -87, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049254d3e63053e20d
2021-11-24 16:15:24 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -88, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049454d3e63053e20d
2021-11-24 16:15:26 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -88, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049554d3e63053e20d
2021-11-24 16:15:28 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -71, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049654d3e63053e20d
2021-11-24 16:15:32 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -58, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049854d3e63053e20d
2021-11-24 16:15:34 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -59, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049954d3e63053e20d
2021-11-24 16:15:36 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -58, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049a54d3e63053e20d
2021-11-24 16:15:38 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -63, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049b54d3e63053e20d
2021-11-24 16:15:40 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -67, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049c54d3e63053e20d
2021-11-24 16:15:42 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -64, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049d54d3e63053e20d
2021-11-24 16:15:44 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -65, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049e54d3e63053e20d
2021-11-24 16:15:46 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -60, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e1049f54d3e63053e20d
2021-11-24 16:15:48 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -61, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104a054d3e63053e20d
2021-11-24 16:15:50 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -56, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104a154d3e63053e20d
2021-11-24 16:15:52 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -54, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104a254d3e63053e20d
2021-11-24 16:15:54 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -64, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104a354d3e63053e20d
2021-11-24 16:15:58 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -72, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104a554d3e63053e20d
2021-11-24 16:16:04 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -62, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104a854d3e63053e20d
2021-11-24 16:16:06 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -67, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104a954d3e63053e20d
2021-11-24 16:16:08 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -59, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104aa54d3e63053e20d
2021-11-24 16:16:10 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -62, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104ab54d3e63053e20d
2021-11-24 16:16:14 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -69, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104ad54d3e63053e20d
2021-11-24 16:16:16 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -69, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104ae54d3e63053e20d
2021-11-24 16:16:55 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -68, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104af54d3e63053e20d
2021-11-24 16:16:57 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -67, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b054d3e63053e20d
2021-11-24 16:16:59 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -67, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b154d3e63053e20d
2021-11-24 16:17:01 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -69, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b254d3e63053e20d
2021-11-24 16:17:03 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -61, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b354d3e63053e20d
2021-11-24 16:17:05 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -61, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b454d3e63053e20d
2021-11-24 16:17:07 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -64, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b554d3e63053e20d
2021-11-24 16:17:09 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -76, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b654d3e63053e20d
2021-11-24 16:17:11 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -63, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b754d3e63053e20d
2021-11-24 16:17:13 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -66, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b854d3e63053e20d
2021-11-24 16:17:15 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -77, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104b954d3e63053e20d
2021-11-24 16:17:21 INFO (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] BLE ADV from UNKNOWN Xiaomi device: RSSI: -84, MAC: E2:53:30:E6:D3:54, ADV: 0f1695fe3030e104bc54d3e63053e20d
```

**Comment by Ernst79** [comment]
hmm, unfortunately no difference, all `3030` messages, which contain no data. Perhaps it needs to be added to Mi Home?

**Comment by andrewjswan** [comment]
The cube was added, and when I rotated its edges, it simultaneously showed
on the phone screen that I was rotating them.

**Comment by Ernst79** [comment]
Hmmm, from what I understand from the xiaomi gateway integration and the other links, that the cube uses communication with a connection (not supported) but also some passive advertisements for rotation. But I don’t see any of these messages In your case. I’m afraid that it can’t be added if that is the case. I’ll leave the issue open in case someone else want to try to get some data from his cube.

**Comment by andrewjswan** [comment]
Thanks!

**Comment by andrewjswan** [comment]
Turns out it needs to be activated for control, 4 quick turns of the edge.
Thanks to @AlexxIT for the tip.
I put the Beta, it appeared in the devices, but no matter how I twist it, it always shows a right turn.
@AlexxIT Gave a link to the code, and said that the Cube sends only a turn to the left **1** and right **0**. Nothing else.
Link: https://github.com/AlexxIT/XiaomiGateway3/blob/45338dd6ca673bfafa61a0bd4b695e08b2b26b96/custom_components/xiaomi_gateway3/core/converters/mibeacon.py#L111

![image](https://user-images.githubusercontent.com/3609660/143455234-a9a37106-481d-433f-ad6b-54f738bac017.png)

**Comment by Ernst79** [comment]
Ah, great. We are making progress. I already had 1 = left, 0 = right, I thought, but I apparently I have taken the wrong byte, I think. 

Could you add the last line in the code below (`_LOGGER.error......`) in `custom_components/ble_monitor/ble_parser/xiaomi.py` on line 239 and do a restart. Afterwards, make some turns with the cube and collect the error messages in the HA log. 

```python
def obj1001(xobj, device_type):
    if len(xobj) == 3:
        (button_type, value, press) = BUTTON_STRUCT.unpack(xobj)
        _LOGGER.error("cube object button_type: %s, value: %s, press %s", button_type, value, press)
```

**Comment by andrewjswan** [comment]
> Afterwards, make some turns with the cube and collect the error messages in the HA log.

```
2021-11-26 11:05:30 ERROR (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] cube object button_type: 0, value: 0, press 0
2021-11-26 11:05:33 ERROR (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] cube object button_type: 0, value: 0, press 0
2021-11-26 11:05:34 ERROR (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] cube object button_type: 0, value: 0, press 0
2021-11-26 11:05:36 ERROR (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] cube object button_type: 0, value: 0, press 0
2021-11-26 11:05:38 ERROR (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] cube object button_type: 1, value: 0, press 0
2021-11-26 11:05:39 ERROR (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] cube object button_type: 1, value: 0, press 0
2021-11-26 11:05:41 ERROR (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] cube object button_type: 1, value: 0, press 0
2021-11-26 11:05:42 ERROR (Thread-5) [custom_components.ble_monitor.ble_parser.xiaomi] cube object button_type: 1, value: 0, press 0
```
0 and 1 seem to coincide with my rotation. 
And you can remove the battery sensor for the cube, it doesn't seem to transmit it anyway... Or add a setting to it, whether to create the battery sensor or not.

**Comment by Ernst79** [comment]
Thanks, I have used `press` in stead of `button_type`. I'll change it this weekend. Will remove the battery sensor as well.

**Comment by Ernst79** [comment]
Should be fixed in 6.1.1-beta. Let me know if it works.

**Comment by andrewjswan** [comment]
Work fine! Thanks!
